src/library/scala/collection/immutable/Vector.scala

```diff
@@ -26,1284 +33,2073 @@ import scala.runtime.Statics.releaseFence
+
+  override final def last: A = {
+    if(this.isInstanceOf[BigVector[_]]) {
+      val v = this.asInstanceOf[BigVector[_]]
+      v.suffix1(v.suffix1.length-1)
+    } else prefix1(prefix1.length-1)
+  }.asInstanceOf[A]
+
+  override final def foreach[U](f: A => U): Unit = {
+    val c = vectorSliceCount
+      var i = 0
+      while(i < c) {
+        foreachRec(vectorSliceDim(c, i)-1, vectorSlice(i), f)
+        i += 1
+      }
+  }
+
+  // The following definitions are needed for binary compatibility with ParVector
+  private[collection] def startIndex: Int = 0
+  private[collection] def endIndex: Int = length
+  private[collection] def initIterator[B >: A](s: VectorIterator[B]): Unit =
+    s.it = iterator.asInstanceOf[NewVectorIterator[B]]
+}
+
+
+/** This class only exists because we cannot override `slice` in `Vector` in a binary-compatible way */
+private sealed abstract class VectorImpl[+A](_prefix1: Arr1) extends Vector[A](_prefix1) {
+
+  override final def slice(from: Int, until: Int): Vector[A] = {
+    val lo = mmax(from, 0)
+    val hi = mmin(until, length)
+    val newlen = hi - lo
+    if(newlen == length) this
+    else if(newlen <= 0) Vector0
+    else slice0(lo, hi)
+  }
+}
+
+
+/** Vector with suffix and length fields; all Vector subclasses except Vector1 extend this */
+private sealed abstract class BigVector[+A](_prefix1: Arr1, private[immutable] val suffix1: Arr1, private[immutable] val length0: Int) extends VectorImpl[A](_prefix1) {
+
+  protected[immutable] final def foreachRest[U](f: A => U): Unit = {
+    val c = vectorSliceCount
+    var i = 1
+    while(i < c) {
+      foreachRec(vectorSliceDim(c, i)-1, vectorSlice(i), f)
+      i += 1
     }
   }
+}
+
+
+/** Empty vector */
+private object Vector0 extends BigVector[Nothing](empty1, empty1, 0) {
+
+  def apply(index: Int): Nothing = throw ioob(index)
+
+  override def updated[B >: Nothing](index: Int, elem: B): Vector[B] = throw ioob(index)
+
+  override def appended[B >: Nothing](elem: B): Vector[B] = new Vector1(wrap1(elem))
+
+  override def prepended[B >: Nothing](elem: B): Vector[B] = new Vector1(wrap1(elem))
 
-  override def updated[B >: A](index: Int, elem: B): Vector[B] = updateAt(index, elem)
+  override def map[B](f: Nothing => B): Vector[B] = this
 
-  override def take(n: Int): Vector[A] = {
-    if (n <= 0)
-      Vector.empty
-    else if (startIndex < endIndex - n)
-      dropBack0(startIndex + n)
-    else
-      this
+  override def tail: Vector[Nothing] = throw new UnsupportedOperationException("empty.tail")
+
+  override def init: Vector[Nothing] = throw new UnsupportedOperationException("empty.init")
+
+  protected[this] def slice0(lo: Int, hi: Int): Vector[Nothing] = this
+
+  protected[immutable] def vectorSliceCount: Int = 0
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef] = null
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int = 0
+
+  override def equals(o: Any): Boolean = {
```

Comment:
I assume this is for performance, but it would only be used in `empty == something`, not in `something == empty`? Would id make sense to override `equals` in `Vector` instead?

Reply:
I didn't put any time into optimizing this further because the old Vector doesn't, either. You could probably gain a bit of performance but `super.equals` doesn't do anything too bad (in particular, it doesn't create any Iterators when both sides have non-matching known sizes).